core/solver/batch_bicgstab_kernels.hpp

```diff
@@ -90,6 +90,103 @@ inline int local_memory_requirement(const int num_rows, const int num_rhs)
+    int n_global;
+    // global stride from one batch entry to the next
+    int gmem_stride_bytes;
+    // padded vector length
+    int padded_vec_len;
+};
+
+
+template <int align_bytes>
+void set_gmem_stride_bytes(storage_config& sconf,
+                           const int multi_vector_size_bytes,
+                           const int prec_storage_bytes)
+{
+    int gmem_stride = sconf.n_global * multi_vector_size_bytes;
+    if (!sconf.prec_shared) {
+        gmem_stride += prec_storage_bytes;
+    }
+    // align global memory chunks
+    sconf.gmem_stride_bytes =
+        gmem_stride > 0 ? ((gmem_stride - 1) / align_bytes + 1) * align_bytes
+                        : 0;
+}
+
+
+/**
+ * Calculates the amount of in-solver storage needed by batch-Bicgstab and
+ * the split between shared and global memory.
+ *
+ * The calculation includes multivectors for
+ * - r
+ * - r_hat
+ * - p
+ * - p_hat
+ * - v
+ * - s
+ * - s_hat
+ * - t
+ * - x
+ * In addition, small arrays are needed for
+ * - rho_old
+ * - rho_new
+ * - omega
+ * - alpha
+ * - temp
+ * - rhs_norms
+ * - res_norms
+ *
+ * @param shared_mem_per_blk  The amount of shared memory per block to use for
+ *   keeping intermediate vectors. In case keeping the matrix in L1 cache etc.
+ *   should be prioritized, the cache configuration must be updated separately
+ *   and the needed space should be subtracted before passing to this
+ *   function.
+ * @param num_rows  Size of the matrix.
+ * @param num_nz  Number of nonzeros in the matrix
+ * @param num_rhs  Number of right-hand-sides in the vectors.
+ * @return  A struct containing allocation information specific to Bicgstab.
+ */
+template <typename Prectype, typename ValueType, int align_bytes = 32>
+storage_config compute_shared_storage(const int shared_mem_per_blk,
+                                      const int num_rows, const int num_nz,
+                                      const int num_rhs)
+{
+    using real_type = remove_complex<ValueType>;
+    const int vec_size = num_rows * num_rhs * sizeof(ValueType);
+    const int num_main_vecs = 9;
+    const int prec_storage =
+        Prectype::dynamic_work_size(num_rows, num_nz) * sizeof(ValueType);
+    int rem_shared = shared_mem_per_blk;
+    storage_config sconf{false, 0, num_main_vecs, 0, num_rows};
+    if (rem_shared <= 0) {
+        set_gmem_stride_bytes<align_bytes>(sconf, vec_size, prec_storage);
+        return sconf;
+    }
+    const int initial_vecs_available = rem_shared / vec_size;
+    const int num_vecs_shared = min(initial_vecs_available, num_main_vecs);
+    sconf.n_shared += num_vecs_shared;
+    sconf.n_global -= num_vecs_shared;
+    if (sconf.n_global > 0) {
+        set_gmem_stride_bytes<align_bytes>(sconf, vec_size, prec_storage);
+        return sconf;
+    }
+    rem_shared -= num_vecs_shared * vec_size;
+    if (rem_shared >= prec_storage && prec_storage > 0) {
+        sconf.prec_shared = true;
+        rem_shared -= prec_storage;
```

Comment:
For correctness, I think this is necessry. For example, if someone wants to print out the `rem_shared` at the end of this function.

Reply:
That has nothing to do with correctness, only with a particular debugging style. As mentioned somewhere else, this modification would make more sense to me if it was stored as part of `sconf`.

Reply:
I think ensuring the variables have correct values is useful for self-documenting purposes, even if they are not used later in the scope. I dont think it makes sense to have rem_shared in `sconf`. Maybe when we split it into two parts, one for core and one for kernels as you suggested, it might. But for now, we can leave it as it is.

Reply:
My issue with dead code is that if someone without deep knowledge of this functions reads this, they might think that something is missing. `rem_shared` is modified, but not used, so why is that there? The reader might then come to the conclusion that there is an error in the code.

Reply:
In combination with the documentation inline, I think there should be no ambiguity. I think within the scope the variable should always have the correct value, so I would not necessarily say that it is dead code.

Reply:
The above is just my interpretation of the code, and my reasoning for removing it. But this is nothing I would hold up the PR for.